src/audio/volume/volume.c

```diff
@@ -774,8 +774,35 @@ static int volume_reset(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_3
+static int volume_init_shim(struct processing_module *mod)
+{
+	struct ipc_config_volume *ipc_volume;
+	struct module_data *mod_data = &mod->priv;
+	const struct sof_ipc_comp_volume *volume =
+		(const struct sof_ipc_comp_volume *)mod_data->cfg.init_data;
+
+	if (IPC_TAIL_IS_SIZE_INVALID(*volume))
+		return -EBADMSG;
+
+	ipc_volume = rballoc(0, SOF_MEM_CAPS_RAM, sizeof(*ipc_volume));
+	if (!ipc_volume)
+		return -ENOMEM;
+
+	ipc_volume->channels     = volume->channels;
+	ipc_volume->initial_ramp = volume->initial_ramp;
+	ipc_volume->max_value    = volume->max_value;
+	ipc_volume->min_value    = volume->min_value;
+	ipc_volume->ramp         = volume->ramp;
+
+	mod_data->cfg.init_data = ipc_volume;
+	mod_data->cfg.data      = ipc_volume;
```

Comment:
IIUC you use `cfg.data` for freeing, isn't `rfree()` missing here - similar to asrc? Please also check others

Reply:
No this is correct. Data is set to the raw IPC so it should not be freed.